### DISCUSSION TITLE
What is the design consideration done in the ChData in fprime-gds?

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#2123] [Q&A]
In `ChData` in `fprime-gds` there is the following method:
```    
    def _compute_display_text(self, val_obj, template):
        """
        Returns the display_text for the channel by computing it. This function is defined so as not to clutter __init__()
        but should not be called elsewhere. Use get_display_text() instead.
        Does not depend on self state so as not to be dependent on the order of initialization.
        """
        # This can happen when constructing empty objects (e.g. when listing channels)
        # in which case we just display the description
        if val_obj is None:
            return template.ch_desc
        temp_val = (
            val_obj.val
            if not isinstance(val_obj, (SerializableType, ArrayType))
            else val_obj.formatted_val
        )
      ...
   ```
The output of this method is a single value or a dictionary depending on the type of `val_obj`. If `val_obj` is a basic type, for example `U8` or `F32`, the method return an `int` or a `float`. If val_obj is an array of `U8` the method return a dictionary containing an array but where each value is cast to string. The cast is done inside `SerializableType.formatted_val`. 
   
The question is why the output is different depending on the type of `val_obj`? Should not always be returned output casted to string or always return the original value?

**Comment by thomas-bc** [comment]
I think the intent here is that `SerializableType` and `ArrayType` can themselves have `format` strings associated to them, which are distinct from the Channel's format string. Think:
```fpp
array DelayArray = [3] U32 format "{} ms"
[...]
telemetry latestDelays: DelayArray format "Delays for the last 3 readings are: {}"
```
By retrieving `val_obj.formatted_val`, the display_text becomes `Delays for the last 3 readings are: ["12 ms", "11 ms", "17ms"]`.
If we were to retrieve `val_obj.val` for an ArrayType, the display_text would be: `Delays for the last 3 readings are [12, 11, 17]` which defeats the purpose of having array format strings here.

Now maybe we would want to keep the types intact if the array were to not define a format string? I'm not sure if that's easily doable, since it's noted in the [formatted_value property docs](https://github.com/fprime-community/fprime-tools/blob/eeafd1a37c8d89e3ff6465c34913f431c7901d1c/src/fprime/common/models/serialize/array_type.py#L64) that elements will be casted to a string.

**Reply by SMorettini** [reply]
I check again the code and I realize I was accessing the data with the `get_dict` function but what I should have use it was the `get_val` because I wanted to retrieve the dictionary representing the channel.

Anyway for consistent  I think the `_compute_display_text` should always return a string text with something like:
```
temp_val = (
            str(val_obj.val)
            if not isinstance(val_obj, (SerializableType, ArrayType))
            else val_obj.formatted_val
        )
```
